mm2src/coins/eth/eth_hd_wallet.rs

```diff
@@ -0,0 +1,206 @@
+        params: EnabledCoinBalanceParams,
+        path_to_address: &HDAccountAddressId,
+    ) -> MmResult<HDWalletBalance<Self::BalanceObject>, EnableCoinBalanceError>
+    where
+        XPubExtractor: HDXPubExtractor + Send,
+    {
+        coin_balance::common_impl::enable_hd_wallet(self, hd_wallet, xpub_extractor, params, path_to_address).await
+    }
+
+    async fn scan_for_new_addresses(
+        &self,
+        hd_wallet: &Self::HDWallet,
+        hd_account: &mut HDCoinHDAccount<Self>,
+        address_scanner: &Self::HDAddressScanner,
+        gap_limit: u32,
+    ) -> BalanceResult<Vec<HDAddressBalance<Self::BalanceObject>>> {
+        scan_for_new_addresses_impl(
+            self,
+            hd_wallet,
+            hd_account,
+            address_scanner,
+            Bip44Chain::External,
+            gap_limit,
+        )
+        .await
+    }
+
+    async fn all_known_addresses_balances(
+        &self,
+        hd_account: &HDCoinHDAccount<Self>,
+    ) -> BalanceResult<Vec<HDAddressBalance<Self::BalanceObject>>> {
+        let external_addresses = hd_account
+            .known_addresses_number(Bip44Chain::External)
+            // A UTXO coin should support both [`Bip44Chain::External`] and [`Bip44Chain::Internal`].
+            .mm_err(|e| BalanceError::Internal(e.to_string()))?;
+
+        self.known_addresses_balances_with_ids(hd_account, Bip44Chain::External, 0..external_addresses)
+            .await
+    }
+
+    async fn known_address_balance(&self, address: &HDBalanceAddress<Self>) -> BalanceResult<Self::BalanceObject> {
+        let balance = self
+            .address_balance(*address)
+            .and_then(move |result| Ok(u256_to_big_decimal(result, self.decimals())?))
+            .compat()
+            .await?;
+
+        let coin_balance = CoinBalance {
+            spendable: balance,
+            unspendable: BigDecimal::from(0),
+        };
+
+        let mut balances = CoinBalanceMap::new();
+        balances.insert(self.ticker().to_string(), coin_balance);
+        let token_balances = self.get_tokens_balance_list_for_address(*address).await?;
+        balances.extend(token_balances);
+        Ok(balances)
+    }
+
+    async fn known_addresses_balances(
+        &self,
+        addresses: Vec<HDBalanceAddress<Self>>,
+    ) -> BalanceResult<Vec<(HDBalanceAddress<Self>, Self::BalanceObject)>> {
+        let mut balance_futs = Vec::new();
+        for address in addresses {
+            let fut = async move {
+                let balance = self.known_address_balance(&address).await?;
+                Ok((address, balance))
+            };
+            balance_futs.push(fut);
+        }
+        try_join_all(balance_futs).await
+    }
+
+    async fn prepare_addresses_for_balance_stream_if_enabled(
+        &self,
+        _addresses: HashSet<String>,
+    ) -> MmResult<(), String> {
+        Ok(())
+    }
+}
+
+impl EthCoin {
+    // Todo: make a common implementation for all coins
```

Comment:
If we want to have a common impl, I suggest to create a trait type then and implement it.

Reply:
We already have `CoinWithDerivationMethod` trait that can be used for this https://github.com/KomodoPlatform/komodo-defi-framework/blob/fb1ccd724af5ae2c6ec5fec672883c131d0441a5/mm2src/coins/lp_coins.rs#L3904
I fixed this here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/26cca4beb06c30e8a4534866d6763cd493dc66a3 but renamed it to all_addresses since there is another `my_addresses` method used for utxo transaction history. The two should be merged in the future by making utxo tx history implementation generic across all coins.